packages/flutter/lib/src/widgets/focus_manager.dart

```diff
@@ -1371,25 +1370,64 @@ class FocusManager with DiagnosticableTreeMixin {
   // The node that has requested to have the primary focus, but hasn't been
   // given it yet.
   FocusNode _nextFocus;
+
   // The set of nodes that need to notify their listeners of changes at the next
   // update.
   final Set<FocusNode> _dirtyNodes = <FocusNode>{};
 
+  // The node that has been marked as needing to be unfocused during the next
+  // focus update.
+  FocusNode _unfocusedNode;
+
+  void _markDetached(FocusNode node) {
+    // The node has been removed from the tree, so it no longer needs to be
+    // notified of changes.
+    assert(_focusDebug('Node was detached: $node'));
+    if (_unfocusedNode == node) {
+      _unfocusedNode = null;
+    }
+    if (_primaryFocus == node) {
+      _primaryFocus = null;
+    }
+    _dirtyNodes?.remove(node);
+  }
+
+  void _markPropertiesChanged(FocusNode node) {
+    _markNeedsUpdate();
+    assert(_focusDebug('Properties changed for node $node.'));
+    _dirtyNodes?.add(node);
+  }
+
+  void _markNextFocus(FocusNode node) {
+    if (_primaryFocus == node) {
+      // The caller asked for the current focus to be the next focus, so just
+      // pretend that didn't happen.
+      _nextFocus = null;
+      // If this node is going to be the next focus, then it's not going to be
+      // unfocused unless we call _markUnfocused again, so unset _unfocusedNode.
+      if (_unfocusedNode == node) {
+        _unfocusedNode = null;
+      }
+    } else {
+      _nextFocus = node;
+      _markNeedsUpdate();
+    }
+  }
+
   // Called to indicate that the given node is being unfocused, and that any
   // pending request to be focused should be canceled.
-  void _willUnfocusNode(FocusNode node) {
+  void _markUnfocused(FocusNode node) {
     assert(node != null);
     assert(_focusDebug('Unfocusing node $node'));
     if (_primaryFocus == node || _nextFocus == node) {
-      if (_primaryFocus == node) {
-        _primaryFocus = null;
-      }
       if (_nextFocus == node) {
         _nextFocus = null;
       }
-      _dirtyNodes.add(node);
+      if (_primaryFocus == node) {
+        _unfocusedNode = node;
```

Comment:
What if unfocused is already set to something else? If that can't happen, maybe assert its null before setting it here?

Reply:
Hmm. I *think* it can' only either be null, or whatever `_primaryFocus` is. I'll add an assert and see.